apps/snmptranslate.1
.RC $Header: /nfs/medea/u0/rel5/rcs/Tools/cmusnmp/apps/snmptranslate.1,v 1.2 1992/07/29 18:17:07 djw Exp $
.\* /***********************************************************
.\" 	Copyright 1988, 1989 by Carnegie Mellon University
.\" 
.\"                       All Rights Reserved
.\" 
.\" Permission to use, copy, modify, and distribute this software and its 
.\" documentation for any purpose and without fee is hereby granted, 
.\" provided that the above copyright notice appear in all copies and that
.\" both that copyright notice and this permission notice appear in 
.\" supporting documentation, and that the name of CMU not be
.\" used in advertising or publicity pertaining to distribution of the
.\" software without specific, written prior permission.  
.\" 
.\" CMU DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING
.\" ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS, IN NO EVENT SHALL
.\" CMU BE LIABLE FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR
.\" ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS,
.\" WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION,
.\" ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS
.\" SOFTWARE.
.\" ******************************************************************/
.TH SNMPTRANSLATE 1 "6 Febuary 1992"
.UC 4
.SH NAME
snmptranslate - Translate mib variables
.SH SYNOPSIS
snmptranslate [-D] [-n] variable-name
.SH DESCRIPTION
snmptranslate is an SNMP application translates a variable-name to or from 
numeric and textual format.  
.PP
The 
.I -n
option signals to translate to symbolic form (from numeric form usually).  
Otherwise the programs translates to numeric form.
.PP
The 
.I -D
option causes the concise-mib format DESCRIPTION clause to be printed for the
specified variable name.  If no description is available, "No description" is 
printed.
.PP
.SH "EXAMPLES"
.PP
.B snmptranslate -D -n system

Outputs:
.RS
.ft R
.nf
 .iso.org.dod.internet.mgmt.mib-2.system
 No description
.fi
.ft R
.RE
.PP
.B snmptranslate -D -n system.sysuptime

Outputs:
.RS
.ft R
.nf
 .iso.org.dod.internet.mgmt.mib-2.system.sysUpTime
 Description: "The time (in hundredths of a second) since the
 network management portion of the system was last
 re-initialized."
.fi
.ft R
.RE
.PP
.B snmptranslate -D system.sysuptime

Outputs:
.RS
.ft R
.nf
 .1.3.6.1.2.1.1.3
 Description: "The time (in hundredths of a second) since the
 network management portion of the system was last
 re-initialized."
.fi
.ft R
.RE
.PP
.PP
.SH "SEE ALSO"
variables(5), RFC 1065, RFC 1212, RFC 1213




HISTORY
Version .91:  Changed the ASN type for the types: Counter, Ipaddress, gauge,
timeticks, and opaque.  The SMI defines them as application specific types,
not ASN_INTEGER and ASN_OCTET_STRING.

Version 1.0: Added snmpnetstat, snmptrap, snmptrapd.  Everything runs on a new
API that has an asynchronous interface.  Most old hard-coded limits are either
non-existent or much larger than necessary.  The mib module now gets it's data
by parsing a text file in the same ASN.1 format as is given in RFC1066.

Version 1.1b: <Not documented>

Version 1.2u: Added Concise MIB parsing support to the snmp library
and to the snmptranslate program.  New mib.txt file that is an
almost-direct extract from RFC1213.  Added the snmproute,
snmptranslate, and snmpdelta programs.  Assorted bug fixes.
Integers are now the only signed numbers.  The snmptrap program is
heavily revised-- it now can pass user-specified varbindings in the
trap PDU.  Some nice snmpnetstat additions from Havard Eidnes
(he@idt.unit.no).


apps/snmptranslate.c
#ifndef lint
static char rcsid[]="$Header: /nfs/medea/u0/rel5/rcs/Tools/cmusnmp/apps/snmptranslate.c,v 1.2 1992/07/29 18:17:09 djw Exp $";
#endif

/*
 * snmptranslate.c - translate mib names
 *
 */
/***********************************************************
	Copyright 1988, 1989 by Carnegie Mellon University

                      All Rights Reserved

Permission to use, copy, modify, and distribute this software and its 
documentation for any purpose and without fee is hereby granted, 
provided that the above copyright notice appear in all copies and that
both that copyright notice and this permission notice appear in 
supporting documentation, and that the name of CMU not be
used in advertising or publicity pertaining to distribution of the
software without specific, written prior permission.  

CMU DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING
ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS, IN NO EVENT SHALL
CMU BE LIABLE FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR
ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS,
WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION,
ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS
SOFTWARE.
******************************************************************/
#include <sys/types.h>
#include <netinet/in.h>
#include <stdio.h>
#include <ctype.h>

#include "snmp.h"
#include "asn1.h"
#include "snmp_impl.h"
#include "snmp_api.h"
#include "snmp_client.h"

extern int  errno;
int	snmp_dump_packet = 0;

main(argc, argv)
    int	    argc;
    char    *argv[];
{
    int	arg, count;
    char *current_name = NULL;
    oid name[MAX_NAME_LEN];
    int name_length;
    int tosymbolic = 0;
    int dumpdescription = 0;
    
    init_mib();
    /*
     * usage: snmptranslate name
     */
    for(arg = 1; arg < argc; arg++){
	if (argv[arg][0] == '-'){
	    switch(argv[arg][1]){
	      case 'n':
		tosymbolic = 1;
		break;	     
	      case 'D':
		dumpdescription = 1;
		break;	     
	      default:
		printf("invalid option: -%c\n", argv[arg][1]);
		break;
	    }
	    continue;
	}
	current_name = argv[arg];
    }
    
    if (current_name == NULL){
	printf("usage: snmptranslate [-n] [-D] object-identifier\n");
	exit(1);
    }
    name_length = MAX_NAME_LEN;
    if (!read_objid(current_name, name, &name_length)){
	printf("Invalid object identifier: %s\n", current_name);
    }
    if (tosymbolic){
	print_objid(name, name_length);
    } else {
	for(count = 0; count < name_length; count++)
	    printf(".%d", name[count]);
	printf("\n");
    }
    if (dumpdescription) {
	print_description(name, name_length);
    }	
    exit(0);
}

